build-logic-settings/environment/src/main/kotlin/gradlebuild.build-environment.settings.gradle.kts

```diff
@@ -18,8 +14,13 @@ import gradlebuild.basics.parentOrRoot
  * limitations under the License.
  */
 
+import gradlebuild.basics.BuildEnvironmentExtension
+import gradlebuild.basics.git
 
-val buildEnvironmentExtension = extensions.create("buildEnvironment", BuildEnvironmentExtension::class)
-buildEnvironmentExtension.gitCommitId = git("rev-parse", "HEAD")
-buildEnvironmentExtension.gitBranch = git("rev-parse", "--abbrev-ref", "HEAD")
-buildEnvironmentExtension.repoRoot = layout.projectDirectory.parentOrRoot()
+val buildLayout = layout
+gradle.lifecycle.beforeProject {
```

Comment:
🤔 Before `git` would be executed only twice per build, right? And now it's once per project?

Reply:
Interestingly it's seems it's called 3-times in both cases during sync, at least when trying to do some println in the zip part I get for both:
```
> Configure project :
Executed 'git rev-parse --abbrev-ref HEAD' for ':'
Executed 'git rev-parse --abbrev-ref HEAD' for ':'
Executed 'git rev-parse --abbrev-ref HEAD' for ':'
```

So it looks like provider is evaulated anytime `git()` is called anyway, or that should not be the case?

If you have any ideas how to improve that, let me know. 
We can also just merge improvement for `gradle/shared-with-buildSrc/mirrors.settings.gradle.kts` that is trivial and we leave this one for some other time when we'll have more tools.

Reply:
Any thoughts if that could be a problem or not @gradle/bt-developer-productivity?

Reply:
It would be possible to always execute `git` once in the settings plugin and then simply propagate the data to every project as a Kotlin data class or similar construct (until we have a proper solution via the _Shared Data API_):

```kotlin
captureBuildEnvironment().let { buildEnvironment ->
  gradle.lifecycle.beforeProject {
     extensions.add("buildEnvironment", buildEnvironment)
  }
}
```

Reply:
How will that affect configuration cache? E.g. will that git read be an input to configuration cache?

Reply:
In my example, yes. Is the git result currently directly wired as a task input or is it ever used at configuration time?

Reply:
I think in some cases it's still used at configuration time, but there was a PR that improved that https://github.com/gradle/gradle/pull/27954.

I was thinking of maybe trying to use Build service. So in root we would register it. And that build service would then cache the result of a git command, so it could be reused in other projects.

Reply:
I think solution with BuildService could work. I tested and cc is reused if commit id changes, at least for integration tests.

And if https://github.com/gradle/gradle/blob/593ea221e7326119aa1c609c58cbf098b081e870/subprojects/core-api/src/main/java/org/gradle/process/ExecOutput.java#L33-L34 is true, than git should be executed just once